Initial commit

clpe04 opened the pull request.

Commit:
Initial commit

jacobemcken commented on README.md:
```diff
@@ -2,6 +2,119 @@
+;; Create a new node on the connected server and return it
+(client/create-node
+  connection
+  {:reference-id "N"
```
Should be `:ref-id`.

jacobemcken commented on src/full_spectrum/neo4clj/internal/neo4j.clj:
```diff
@@ -0,0 +1,167 @@
+  (let [logging-level (ConsoleLogging. (condp = (get-in [:logging :level] opts)
+                                         :all     Level/ALL
+                                         :error   Level/SEVERE
+                                         :warning Level/WARNING
```
If you decide the keywords then I would suggest `:warning` to be renamed to `:warn` so the length matches the rest of the keywords (just like information is shortened to `:info`).

clpe04 replied:
Makes sense

jacobemcken commented on src/full_spectrum/neo4clj/internal/neo4j.clj:
```diff
@@ -0,0 +1,167 @@
+
+(defmulti neo4j->clj
+  "Converts a Neo4J internal entity to a Clojure Hash-Map"
+  (fn [entity] (class entity)))
```
I should be enough to just do the follow:
```clojure
(defmulti neo4j->clj
  "Converts a Neo4J internal entity to a Clojure Hash-Map"
  class)
```

jacobemcken commented on README.md:
```diff
@@ -2,6 +2,119 @@
+  connection
+  {:reference-id "N"
+   :labels [:person]
+   :properties {:first-name "Tim"
```
The name should be Thomas Anderson, as a reference to The Matrix.
Neo's name when he was born in the Matrix.

https://en.wikipedia.org/wiki/Neo_(The_Matrix)

clpe04 replied:
I like that :smile:

jacobemcken commented on README.md:
```diff
@@ -2,6 +2,119 @@
+
+~~~~
+{:id       34                            ;; The id from Neo4J, this is only set if object is fetched from Neo4J
+ :ref-id   "G_123"                       ;; The variable name used for this entity in bolt queries
```
Gensyms use double underscore if I recall correctly (`G__123`).

clpe04 replied:
You're correct

jacobemcken commented on README.md:
```diff
@@ -2,6 +2,119 @@
+### By id
+
+~~~~
+{:ref-id    "G312"                        ;; The variable name used for this entity in bolt queries
```
Format should at least match the other reference ids, whether you decide to use single or double underscore.

jacobemcken commented on README.md:
```diff
@@ -2,6 +2,119 @@
 
+# Getting started
+
+~~~
```
You should be able to supply the type of code. I know it works with `` ` `` not


```
~~~clojure
;; some Clojure code
(defn hey [] bla)
~~~
```

Will result in:
~~~clojure
;; some Clojure code
(defn hey [] bla)
~~~

jacobemcken commented on README.md:
```diff
@@ -2,6 +2,119 @@
+
+## Supported Neo4J Server Versions
+
+### Neocons 3.2
```
Does this belong here? I thought the entire purpose of this lib was to avoid an extra unnecessary dependency (Neocons)?

jacobemcken commented on project.clj:
```diff
@@ -0,0 +1,13 @@
+                 [com.rpl/specter "1.1.2"]
+                 [org.clojure/clojure "1.10.0"]
+                 [org.neo4j.driver/neo4j-java-driver "1.7.2"]]
+  :main full-spectrum.neo4clj.client
```
Not used for anything.

jacobemcken commented on project.clj:
```diff
@@ -0,0 +1,13 @@
+                 [org.clojure/clojure "1.10.0"]
+                 [org.neo4j.driver/neo4j-java-driver "1.7.2"]]
+  :main full-spectrum.neo4clj.client
+  :aot [full-spectrum.neo4clj.client]
```
I think this is only needed if you want to interface with this lib from Java ?!? :thinking: 
I think you can safely delete it but test to be sure.

jacobemcken commented on project.clj:
```diff
@@ -0,0 +1,13 @@
+  :profiles {:dev {:dependencies [[org.clojure/tools.namespace "0.2.11"]
+                                  [criterium "0.4.4"]]
+                   :source-paths ["dev"]
+                   :repl-options {:init-ns user}}})
```
Isn't the `user` namespace usually the default REPL namespace. Read: "Delete unless it adds value to the project in any way".

jacobemcken commented on src/full_spectrum/neo4clj/client.clj:
```diff
@@ -0,0 +1,118 @@
+  "Execute the given queries on the specified connection within the same session"
+  [conn & queries]
+  (with-open [session (neo4j/create-session conn)]
+    (for [query queries]
```
I don't know if it is a good idea, but have you considered if `(doall (mapv ... ` would be a good alternative for multi threading the queries.

jacobemcken commented on src/full_spectrum/neo4clj/client.clj:
```diff
@@ -0,0 +1,118 @@
+        (neo4j/rollback-transaction trans)
+        (throw e)))))
+
+(defmacro execute-in-transaction2
```
This looks like some development leftovers because of the post pended  `2`.

jacobemcken commented on src/full_spectrum/neo4clj/client.clj:
```diff
@@ -0,0 +1,118 @@
+  (get (first (execute! conn (builder/create-node-query node true)))
+       (:ref-id node)))
+
+(defn find-nodes!
```
All the exclamation marks seems a bit excessive. I thought Clojure only used those when mutating one or more of the arguments given to the function. 

I.e. `spit` has side effects and writes to a file (mutating the underlying filesystem :grin: ), but isn't named `spit!`.

clpe04 replied:
Based on the official JDBC Clojure ns it seems like it is used for all calls mutating the DB, so find-nodes should not have and exclamaton mark, but create, update and delete should

jacobemcken replied:
You have 2x "should". I expected one of them to be "should not"?

jacobemcken commented on src/full_spectrum/neo4clj/internal/neo4j.clj:
```diff
@@ -0,0 +1,167 @@
+       (map (fn [^Record r] (.asMap r)))
+       (map #(reduce-kv (fn [m k v] (assoc m k (neo4j->clj v))) {} %))))
+
+#_(map #(reduce (fn [m [k v]] (assoc m k (neo4j->clj v))) {} %)
```
Code leftovers that needs to be removed?

jacobemcken commented on project.clj:
```diff
@@ -0,0 +1,13 @@
+(defproject neo4clj "1.0.0-SNAPSHOT"
```
I think the following is possible ?!?

Maybe give the project a group like `fullspectrum/neo4clj` but then remove the `full-spectrum.` from all namespaces because it is now possible to specify in the dependency instead:

```
;; projects.clj
:dependencies [[fullspectrum/neo4clj "1.0.0"]]

;; Any clojure file
(:require [neo4j.client :as client])
```

Btw. is it on purpose that you didn't name it neo4jclj (2 j's)?

clpe04 replied:
Ill test it out, and yes it is on purpose it is named without the second j

jacobemcken reviewed (commented):
Really REALLY nice work... looks very good1 :+1:

jacobemcken commented on src/full_spectrum/neo4clj/client.clj:
```diff
@@ -0,0 +1,118 @@
+
+(defn add-labels!
+  "Takes a collection of labels and adds them to the found neo4j entities"
+  [^StatementRunner conn ^clojure.lang.APersistentMap neo4j-entity ^clojure.lang.APersistentVector labels]
```
Consider if it increases readability putting these on multiple lines for all functions with "long" argument declerations:
```clojure
[^StatementRunner conn
 ^clojure.lang.APersistentMap neo4j-entity
 ^clojure.lang.APersistentVector labels]
```

jacobemcken commented on src/full_spectrum/neo4clj/internal/neo4j.clj:
```diff
@@ -0,0 +1,167 @@
+(ns full-spectrum.neo4clj.internal.neo4j
```
My personal preference is to drop `.internal` from the namespace(s).

It doesn't make it anymore internal having that as part of the name. Just don't mention them in documentation. Something that IMO would add more value is to write what your intention with an "internal" namespace is in the `ns` documentation:

```clojure
(ns neo4clj.neo4j
  "This namespace is intended for internal use of the library and is subject to
change. Don't expect function compatibility across version updates."
```

Also maybe the name of the NS could be improved. `...neo4j` doesn't give much away about what to expect to find in there :-)

The first that catches my attention when scanning through the NS is that several of the functions are converting so maybe it would make sense to split at least those into their own "convert/transform/whatnot" namespace.

jacobemcken commented on README.md:
```diff
@@ -2,6 +2,119 @@
 
 Neo4clj is a idomatic clojure client.
```
Change to the following for a better "sell":
> ... clojure client, exclusivly using [Bolt][] for performance.

[Bolt]: https://boltprotocol.org/

jacobemcken committed:
Improve README with code highlight & other minor things

I removed IVS because it is very Danish and branding-wise it might just confuse.
I suggest that we put a link to our website with all the borring info like IVS
and CVR etc.

jacobemcken committed:
Use easy to extend version layout

A table is much easier to extend than adding new headings like Neocons does.

jacobemcken committed:
Cleanup unused parts in project.clj

This also fixed a bug causing `lein repl` to throw an error:

    Could not locate user__init.class, user.clj or user.cljc on classpath.

jacobemcken committed:
Ease reability by using multi line function arg lists

This is only relevant for those functions that har a "long" argument definition
usually due to type hinting.

jacobemcken commented on src/full_spectrum/neo4clj/internal/neo4j.clj:
```diff
@@ -0,0 +1,167 @@
+  :logging :level [:all :error :warning :info :off] - defaults to :warning
+  :encryption [:required :off] - defaults to :required"
+  ^Config [opts]
+  (let [logging-level (ConsoleLogging. (condp = (get-in [:logging :level] opts)
```
I think there is a bug here... the map is supposed to be the first argument to `get-in` unless `condp` is doing some magic.

jacobemcken commented on src/full_spectrum/neo4clj/internal/neo4j.clj:
```diff
@@ -0,0 +1,167 @@
+
+  Supports the current options:
+  :logging :level [:all :error :warning :info :off] - defaults to :warning
+  :encryption [:required :off] - defaults to :required"
```
I would suggest either to stick to the naming from the driver (`:required` / `:none`) or go all the way (`:on` / `:off`). I think a mix is confusing.

jacobemcken committed:
Simplify build-config

Notice that the options map includes renames:
:logging -> :log
:off -> :none

jacobemcken committed:
Cleanup unnecessary code in internal neo4j NS

jacobemcken committed:
Encapsulate data convertion in its own namespace

jacobemcken committed:
Reduce Full Spectrum references to package name

It should be possible to test locally using `lein install`.

jacobemcken committed:
Add CircleCI config for unit test

jacobemcken committed:
Ignore irrelevant files for Git tracking

Commit:
Move logic from internal ns to client ns

  - Also renamed internal neo4j ns to java-interop

Commit:
Add acknowledgements to the README file

Commit:
Add sanitize functions for partameter maps used in CYPHER queries

Commit:
Improve readability of create-relationship-query

Commit:
Set license for the project to MIT

Commit:
Rewrite query build methods to be more reuseable

Commit:
Change interfaces for convert dispatch

  - Also added tests for the convert ns

Commit:
Move files from internal to root directory

jacobemcken commented on src/neo4clj/convert.clj:
```diff
@@ -0,0 +1,84 @@
+  (when m
+    (reduce-kv (fn [m k v] (assoc m (sanitize/cypher-property-key k) (clj-value->neo4j-value v))) {} m)))
+
+(defn clj-node->neo4j
```
This function doesn't seem to be used... should it be deleted?

clpe04 replied:
Removed

jacobemcken committed:
Refactor generating Cypher labels

jacobemcken committed:
Remove unused convertion code

jacobemcken committed:
Create namespace for Cypher helper functions

jacobemcken committed:
Categorize generate-ref-id as Cypher helper

jacobemcken committed:
Move Cypher properties helper to propper namespace

jacobemcken committed:
Move Cypher "where" helper fn to propper namespace

jacobemcken committed:
Move Cypher "node" helper to propper namespace

jacobemcken committed:
Move Cypher "relationship" helper fn to propper ns

Commit:
Changed signature of the connect function

Commit:
Add support for multi key indexes

Commit:
Rename convert function for relatiohsips

Commit:
Add get-graph function

Commit:
Add support for parameters on execute! command

Commit:
Update documentation

  - Added separate docs for Clojure representations and examples

Commit:
Fix wrong function name in query builder namespace

Commit:
Fix old function use in query-builder ns

Commit:
Remove uneeded converts from get-graph

Commit:
Fix wrong test case for index queries

Commit:
Fix wrong expected cyoher query for create-index builder

Commit:
Update links in the example.md file

Commit:
Add new cyoher lookup representation

  - Remove lookup functionality from node representation

Commit:
Renamed transaction methods

Commit:
Added missing examples an ToC

Commit:
Minor cleanup in cypher builders

Commit:
Fix typos and errors in examples

Commit:
Updated projects based on review

jacobemcken commented on docs/examples.md:
```diff
@@ -40,6 +41,20 @@ Below we have given examples of the most common operations in Neo4clj.
+~~~clojure
+(def conn (client/connect "bolt://localhost:7687"))
+
+(def conn (client/connect "bolt://localhost:7687"))
```
This line is a duplicate of the one just above.

clpe04 replied:
Fixed

jacobemcken commented on docs/examples.md:
```diff
@@ -40,6 +41,20 @@ Below we have given examples of the most common operations in Neo4clj.
 
+## Basic requirements
+
+All examples will assume you have added Neo4clj to your project dependencies and required
```
I don't think the following is necessary:
> assume you have added Neo4clj to your project dependencies

Here is my take on the section:

To make the client available and setup a connection run the following:
```clojure
(require '[neo4clj.client :as client])

(def conn (client/connect "bolt://localhost:7687"))
```

Make the Neo4j client and connection available in the `client` and `conn` symbols respectively **is a requirement for the subsequent examples to work.** You might need to adjust the connection parameters to match your setup, see below.

clpe04 replied:
Rewritten with your proposed text

jacobemcken commented on docs/examples.md:
```diff
@@ -0,0 +1,363 @@
+
+## Update and Delete Neo4j Entity
+
+In this section you can find examples on how to update specific parts a Node or Relationship and how to delete them.
```
I think an introduction to the term "entity" would be nice.

I was thinking something along the lines:
> Nodes and relationships are both "entities" and this section will describe how to update and delete them.

jacobemcken commented on docs/examples.md:
```diff
@@ -0,0 +1,363 @@
+### Update Node or Relationship (properties)
+
+To change the properties of a node or relationship we have added two convenience functions.
+Both the `update-properties!` and `replace-properties!` functions takes a lookup representation as second argument.
```
Function names not aligned with new naming.

clpe04 replied:
Fixed

jacobemcken commented on docs/examples.md:
```diff
@@ -0,0 +1,363 @@
+
+This function will update an existing properties map based on the following rules.
+
+⋅⋅* Keys existing only in the given property map is added to the object
```
This looks weird when markdown is rendered. Maybe delete the `..`?

Also it isn't clear what "the given property maps" is?

I'm assuming that it is the third argument? If the lookup representation gets documented in the text above:
> ... takes a lookup representation as second argument.

Why not do the same for "the given property map"?


I don't know if the following would be correct?:
> To change the properties of a node or relationship we have added two convenience functions.
Both will take a [lookup representation](representations.md) as second argument and a map with the properties that should be updated as the third.

clpe04 replied:
I had used MD for unordered sublist instead of just unordered list, fixed now. I have also added additional description to the function.

jacobemcken commented on docs/examples.md:
```diff
@@ -0,0 +1,363 @@
+
+To make it easier to create nodes and relationships, we have made a function to do it in one single call.
+
+The `lookups` key allows you to lookup and refer existing nodes in the Neo4j database. The entries are lookup representations.
```
Here (I assume) the word "entries" refers to an item in the collection (vector)? I is a bit confusing that we are using the word `Entity` for a common term covering both nodes and relationships but now it is being used to cover something else. Or maybe I'm wrong?

clpe04 replied:
This have been removed in the rewrite of the descriptions

jacobemcken commented on docs/examples.md:
```diff
@@ -0,0 +1,363 @@
+The `relatioships` key dictates what to create. The entries are relationship representations and it's possible to use lookups ref-id in the `to` and `from` keys.
+The `returns` key dictates which entities to return as the result of the call. This is a vector of ref-id from collections in the other three keys.
+
+To learn more about the Clojure representation of a create graph structure and it's individual parts please see our [representations](representations.md) page
```
I think I would delete this line and just embed the link to `representations.md` in the above text with direct links:

    The `lookups` key specifies a list (vector) of [lookup representations](representations.md#lookup) referring existing nodes in the database.

clpe04 replied:
Good idea I have updated both graph examples

jacobemcken commented on docs/representations.md:
```diff
@@ -0,0 +1,113 @@
+The `ref-id` is used to identify a entity in a cypher query and allows the use of the same entity in other parts of the query.
+
+Meanwhile the `id` is used to represent an actual entity in the Neo4j database and can be used in fetches to lookup specific entities.
+The `id` will always be se when an entity is fetched from the Neo4j database.
```
Typo in `be se when`

jacobemcken reviewed (commented):
Awesome job in "Clojure representations" :+1: 
Some of the text in examples still seems a bit rough, but overall okay and definitely GOOD ENOUGH for a first release :-)

Commit:
Update project settings

Commit:
Clarified several examples

Commit:
Style the argument description for graphs
